src/base/snark0.ml

```diff
@@ -57,26 +57,22 @@ struct
 
   module Data_spec = struct
     type ('r_var, 'r_value, 'k_var, 'k_value) t =
-      | ( :: ) :
+      | Data_spec :
           ( 'var
           , 'value
           , field
           , (unit, field) Checked.Types.Checked.t )
           Types.Typ.typ
-          * ('r_var, 'r_value, 'k_var, 'k_value) t
-          -> ('r_var, 'r_value, 'var -> 'k_var, 'value -> 'k_value) t
-      | [] : ('r_var, 'r_value, 'r_var, 'r_value) t
+          -> ('r_var, 'r_value, 'var -> 'r_var, 'value -> 'r_value) t
 
     let size t =
-      let rec go :
+      let go :
           type r_var r_value k_var k_value.
           int -> (r_var, r_value, k_var, k_value) t -> int =
        fun acc t ->
         match t with
-        | [] ->
-            acc
-        | Typ { size_in_field_elements; _ } :: t' ->
-            go (acc + size_in_field_elements) t'
+        | Data_spec (Typ { size_in_field_elements; _ }) ->
+            acc + size_in_field_elements
       in
       go 0 t
```

Comment:
```ocaml
let size :
          type r_var r_value k_var k_value.
          int -> (r_var, r_value, k_var, k_value) t -> int =
       function 
        | Data_spec (Typ { size_in_field_elements; _ }) ->
            size_in_field_elements
```

^ simpler no?

Reply:
(I think my comment is moot, I glanced and you seem to remove that altogether later)